design/resource-filter-field-selectors-design.md

```diff
@@ -0,0 +1,158 @@
+* Dynamically generated custom resource definitions
+
+## Goals
+
+- Users can define backup filter criteria based on Kubernetes field selectors.
+
+## Non Goals
+
+- N/A
+
+## High-Level Design
+
+Extends the `ResourcePolicies` API with a `FieldSelectorPolicies` configuration.
+
+## Detailed Design
+
+The `FieldSelectorPolicies` works in complement to the `IncludeClusterScopedResources` API.
+For example, to include the Nginx ingress controller CRD, one can define a field selector policy that looks like this:
+
+```sh
+cat <<EOF > backup-policy-nginx-crd.yaml
+version: v1
+fieldSelectorPolicies:
+- groupResource: 
+    group: apiextensions.k8s.io
+    resource: customresourcedefinitions
+  metadata.name: virtualservers.k8s.nginx.org
+  metadata.name: virtualserverroutes.k8s.nginx.org
+  metadata.name: transportservers.k8s.nginx.org
+  metadata.name: policies.k8s.nginx.org
+EOF
+```
+
+This YAML policy is used to create the policy configmap used by Velero to backup the CRDs:
+
+```sh
+kubectl -n velero create cm backup-policy-nginx-crd --from-file backup-policy-nginx-crd.yaml
+
+velero backup create --include-cluster-scoped-resources="customresourcedefinitions" --resource-policies-configmap=backup-policy-nginx-crd
+```
+
+Although the supported field selectors vary across Kubernetes resource types, all resource types support the `metadata.name` and `metadata.namespace` fields.
+Using unsupported field selectors produces an error.
+Only the `=` operator is supported by this proposed API.
+For more information, see the Kubernetes documentation on [field selectors][1].
+
+### API Design
+
+The following code snippet illustrates the proposed API structure to encapsulate the field selector properties:
+
+```go
+type ResourcePolicies struct {
+  Version        string                          `yaml:"version"`
+  VolumePolicies []VolumePolicy                  `yaml:"volumePolicies"`
+
+  // FieldSelectorPolicies defines a list of policies to enable Kubernetes
+  // resource field selection to support filtering by resource fields.
+  FieldSelectorPolicies []FieldSelectorPolicy    `yaml:"fieldSelectorPolicies"`
+
+  // we may support other resource policies in the future, and they could be added separately
+  // OtherResourcePolicies: []OtherResourcePolicy
+}
+
+import (
+  k8s.io/apimachinery/pkg/fields
+  metav1 k8s.io/apimachinery/pkg/apis/meta/v1
+)
+
+// FieldSelectorPolicy describes the field selection properties used to identify a
+// Kubernetes resource.
+type FieldSelectorPolicy struct {
+  // GroupResource specifies a group and a resource.
+  GroupResource metav1.GroupResource `json:"groupResource,omitempty"`
+
+  // FieldSelectors specifies a field name and its value, typically, implemented
+  // as a map of string-to-string.
+  // If omitted, then all resources of the group will be selected.
+  // +optional
+  FieldSelectors fields.Set `json:"fieldSelectors,omitempty"`
+}
+```
+
+### Code Design
+
```

Comment:
@ihcsim 
 The API above looks good, and could you please give more details in `code design`, how and where to filter resources when doing backup in Velero

Reply:
Sure thing - I will take a stab at it. I am still unfamiliar with Velero code base, might take slightly longer.